### PR TITLE
[FX] Adds C-level monkeypatching of `torch.randn` so that we can capture it during tracing.

### PR DESCRIPTION
**Pull request by Chillee**
```
def foo(x):
    return x + torch.randn(3, 3)

fx.enable_ctracing(True)
print(fx.symbolic_trace(foo).code)
```
results in
```
def forward(self, x):
    randn = torch.randn(3, 3)
    add = x + randn;  x = randn = None
    return add
```

Seems to slow down tracing by 1.5-3x.

DenseNet121: 0.05 -> 0.12 seconds
ResNet18: 0.10 -> 0.15

### CODE REVIEW

File: torch/fx/symbolic_trace.py [hunk 1]
```diff
@@ -602,6 +637,10 @@ def my_custom_function(x, y):
     _wrapped_fns_to_patch.append((f.f_globals, fn_name))
     return fn_or_name
 
+
+
+
+
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:643]: y tho

File: torch/fx/symbolic_trace.py [hunk 2]
```diff
@@ -48,6 +50,38 @@ def _patch_function(fn: FunctionType, nargs: int) -> FunctionType:
     # we can't call this function normally, otherwise it would try to unpack them
     # instead, let's make python think that args and kwargs are normal variables
 
+class CPatchManager(object):
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:53]: Please add a docstring/comment describing what this is used for

**Review by jamesr66a — COMMENTED** [review]
1. Please mark the PR as [WIP] in the title, since obviously this can't be merged hardcoding `randn` all over the place
2. Please add a test to make sure the `randn` case is covered. This may be why the CI is passing: the C++ portion just isn't being hit

File: torch/csrc/fx/fx_init.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      METH_VARARGS | METH_KEYWORDS,
+      "Replaced method implementation."};
+
+  ToRestore to_restore = {};
+  if (!PyArg_ParseTuple(
+          args, "OO", &to_restore.patched_method, &to_restore.patch_fn)) {
+    return nullptr;
+  }
+  if (!PyCFunction_Check(to_restore.patched_method)) {
+    PyErr_SetString(PyExc_RuntimeError, "not a CFunction");
```
Comment (jamesr66a) [inline torch/csrc/fx/fx_init.cpp:69]: To help with diagnosing bug reports:

```
    std::stringstream err;
    err << "Patched object ";
    PyObject *obj_repr = PyObject_Repr(to_restore.patched_method);
    if (PyUnicode_check(obj_repr)) {
        err << PyUnicode_AS_DATA(obj_repr) << " ";
    }
    << " is not a CFunction. Please report a bug to PyTorch!";
    PyErr_SetString(PyExc_RuntimeError, err.str().c_str());
```

**Review by jamesr66a — COMMENTED** [review]
Looks fine. Please also include performance benchmarks showing how this affects tracing speed in the PR description

File: test/test_fx.py [hunk 4]
```diff
@@ -2155,6 +2156,59 @@ def test_annotation_with_future(self):
...
+        rn(torch.rand(1, 3, 224, 224))
+        print("testing print patch")
+        sys.setprofile(None)
+        assert(cnt != 0)
+
+    def test_randn(self):
+        def f():
+            return torch.randn(3, 3)
+
+        enable_ctracing(False)
```
[the remark this answers is not in the document]
Comment (Chillee) [reply test/test_fx.py:2200]: I was kind of reluctant about increasing the API surface of `symbolic_trace` - it's already kind of awkward to thread things like `concrete_args` through a couple different functions.

Perhaps if we make it an attribute of `Tracer` that'd be better.

File: torch/csrc/fx/fx_init.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+exit:
+  Py_DECREF(to_restore->patched_method);
+  Py_DECREF(to_restore->patch_fn);
+  Py_DECREF(self);
+  Py_DECREF(kwargs);
+  return result;
+}
+
+
+static PyObject* patch_function(PyObject* self, PyObject* args) {
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/csrc/fx/fx_init.cpp:58]: Done.

**Review by jamesr66a — APPROVED** [review]

**Pull request closed without merging**